### DISCUSSION TITLE
[FEATURE REQUEST] 2FA/MFA & Re/HCaptch Options

### DISCUSSION THREAD

**Discussion by tezgno** [filebrowser/filebrowser#1466] [Ideas]
With the rise of spray attacks and threat actors, would it be possible to enable options for 2FA/MFA (such as Google Authenticator TOTP) and Re/HCaptcha? This would allow for me to use File Browser and leave it online as opposed to launching it adhoc as necessary. Overall, this is a great tool, but strengthening the security of the tool would be awesome.

**Comment by cmurrayis** [comment]
2nd this.

**Reply by piraterx** [reply]
I would love something like this.

**Comment by TCB13** [comment]
I've been keeping this issue requesting the same open for a while: https://github.com/filebrowser/filebrowser/issues/1827

How I envision it working:

- Add an option so TOTP 2FA can be enabled for users;
- Generate a QR code for each account;
- Users scan the QR code into their phones with a TOTP compatible App such as Google Authenticator;
- On login users must provide username + password + temporary code.

Meanwhile @tezgno I can recommend you to either use some authentication solutions [such as Authelia](https://github.com/filebrowser/filebrowser/issues/1827#issuecomment-1047777732) or even a simple [Nginx config that provides pre-authentication and 2FA](https://seantodd.co.uk/blog/putting-2fa-on-everything/). To be fair you should[ always run FileBrowser behind a Nginx reverse proxy](https://github.com/filebrowser/filebrowser/issues/400) as it will reduce the attack surface.

Speaking about recommendations I would suggest you to run FileBrowser and make it only accessible through a Wireguard VPN for increased security. [Setup Wireguard VPN](https://www.digitalocean.com/community/tutorials/how-to-set-up-wireguard-on-debian-11) on the server. [There's also this nice UI](https://github.com/ngoduykhanh/wireguard-ui) that can be used to do most of the setup and create client config files. For the VPN port use custom 5 digit number - something like 23901 (up to 65535) to make your service harder to find.

But... as I understand you may want to access FileBrowser from machine where you can't install a VPN client you may want to:

1. In the server consider setting up nftables / iptables / another firewall [10 minute guide](https://wiki.nftables.org/wiki-nftables/index.php/Quick_reference-nftables_in_10_minutes);
2. If you're running this at home with NAT, configure nftables to only allow traffic coming from public IP addresses to the FileBrowser port - this will protect your server if by some mistake the router starts forwarding more traffic from the internet to the server than it should;
3. Configure nftables to restrict what IPs or countries are allowed to access your server. Most likely you only need to allow incoming connection from your office static IP or eventually your entire country (https://wiki.nftables.org/wiki-nftables/index.php/GeoIP_matching).

If you don't want to setup a firewall you can alternatively configure your nginx to [geo-block requests](https://docs.nginx.com/nginx/admin-guide/security-controls/controlling-access-by-geoip/).
